L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,236 @@
+                    grpc_tls_credential_reload_arg* arg),
+    void (*cancel)(void* config_user_data, grpc_tls_credential_reload_arg* arg),
+    void (*destruct)(void* config_user_data));
+
+/** --- TLS server authorization check config. --- */
+typedef struct grpc_tls_server_authorization_check_arg
+    grpc_tls_server_authorization_check_arg;
+
+/** callback function provided by gRPC used to handle the result of server
+    authorization check. It is used when schedule API is implemented
+    asynchronously, and serves to bring the control back to gRPC C core. */
+typedef void (*grpc_tls_on_server_authorization_check_done_cb)(
+    grpc_tls_server_authorization_check_arg* arg);
+
+/** A struct containing all information necessary to schedule/cancel a server
+   authorization check request. cb and cb_user_data represent a gRPC-provided
+   callback and an argument passed to it. result will store the result of
+   server authorization check. target_name is the name of an endpoint the
+   channel is connecting to and certificate represents a complete certificate
+   chain including both signing and leaf certificates. status and error_details
+   contain information about errors occurred when a server authorization check
+   request is scheduled/cancelled. */
+struct grpc_tls_server_authorization_check_arg {
+  grpc_tls_on_server_authorization_check_done_cb cb;
+  void* cb_user_data;
+  int result;
+  const char* target_name;
+  const char* peer_cert;
+  grpc_status_code status;
+  const char* error_details;
+};
+
+/** Create a grpc_tls_server_authorization_check_config instance.
+    - config_user_data is config-specific, read-only user data
+      that works for all channels created with a credential using the config.
+    - schedule is a pointer to an application-provided callback used to invoke
+      server authorization check API. The implementation of this method has to
+      be non-blocking, but can be performed synchronously or asynchronously.
+      1)If processing occurs synchronously, it populates arg->result,
+      arg->status, and arg->error_details and returns zero.
+      2) If processing occurs asynchronously, it returns a non-zero value. The
+      application then invokes arg->cb when processing is completed. Note that
+      arg->cb cannot be invoked before schedule API returns.
+    - cancel is a pointer to an application-provided callback used to cancel a
+      server authorization check request scheduled via an asynchronous schedule
+      API. arg is used to pinpoint an exact check request to be cancelled. The
+      operation may not have any effect if the request has already been
+      processed.
+    - destruct is a pointer to an application-provided callback used to clean up
+      any data associated with the config.
+*/
+GRPCAPI grpc_tls_server_authorization_check_config*
+grpc_tls_server_authorization_check_config_create(
+    const void* config_user_data,
+    int (*schedule)(void* config_user_data,
```

Comment:
note that the pattern  where C core calls a user defined callback, which in turn should call a "done" callback provided as one of the arguments is a pattern that might work well in C and C++, but it is difficult to implement in wrapper languages (it involves a callback from native code back to managed code, which can be troublesome by itself because there might be limitations on what code can run in such constrained environment and with garbage collected languages managing the lifecycle of all the involved objects correctly is hard).

We did something similar with the authentication plugin API in the past and it has resulted in quite a lot of ugly C# code, so perhaps this time we could give more thought to making sure the new API is easily wrappable in other languages than C++.

Reply:
@jtattermusch good point. Are there examples of how to do it right with respect to all the wrapped languages? Or a document that describes how to do it?

Reply:
@markdroth
 
Thanks @jtattermusch for the feedback. When designing async SPIFFE API, we indeed followed the similar pattern as that of plugin credentials. For credential reload, we decided not to support async callback (I will update the proposal) since it turns out not giving too much benefit, but for server authorization check, async callback will be still supported. 

As @sanjaypujare suggested, it will be great if you could provide more details on what the wrapped languages expect w.r.t async callback. It is unclear to me how the current API's needs to be changed to make wrapped languages's implementation easier.

Reply:
The SPIFFE API supports both sync and async callbacks, so that we are flexible for future changes. We can only expose sync callback to python and other wrapped languages, where sync and async callbacks are only available to C++.